llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp

```diff
@@ -3996,6 +4001,8 @@ unsigned X86AsmParser::checkTargetMatchPredicate(MCInst &Inst) {
 
   if (UseApxExtendedReg && !X86II::canUseApxExtendedReg(MCID))
     return Match_Unsupported;
+  if (ForcedNoFlag != static_cast<bool>(MCID.TSFlags & X86II::EVEX_NF))
```

Comment:
Will use `!!`.  No, the logic is 

1. NF instruction can be matched only when {nf} exsits
2.  non-NF instruction can not be matched when {nf} exists

so the suggested code does not work.

Reply:
Why a not `{nf}` assembly can be matched to {nf} instruction? If you want to check for unexpected case, you can use an assert here.

Reply:
It's b/c we don't add `{nf}` in the asm string in TD explicitly

Reply:
Then how can non `{nf}` instruction correctly matched to non {nf} instruction?
Or we should concat `{nf}` in asm string since its mandatory?

Reply:
NF instruction and non-NF instruction shares the same asm string in the TD file.
Fonr non `{nf}` asm, `ForceNoFlag = false`, 
for NF instruction,  `(MCID.TSFlags & X86II::EVEX_NF) = true`.
 
They're not equal. So non `{nf}` asm can only match non-NF instruction.

Reply:
`{evex}` doesn't check in this way. Given a asm string without `{nf}`, it should be selected to non-NF instruction, otherwise, there's something wrong during match instruction. That's why I think it just need to assert for this case.

Reply:
Okay. One usage of `{evex}`: `ExplicitEVEX` checks this way.  Anyway, concat `{nf}` would complicate code in TD a lot.